runtime/src/main/kotlin/kotlin/native/concurrent/MutableData.kt

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright 2010-2018 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the LICENSE file.
+ */
+
+package kotlin.native.concurrent
+
+import kotlin.native.internal.*
+import kotlinx.cinterop.*
+
+@SymbolName("Kotlin_Any_share")
+external private fun Any.share()
+
+@SymbolName("Kotlin_CPointer_CopyMemory")
+external private fun CopyMemory(to: COpaquePointer?, from: COpaquePointer?, count: Int)
+
+/**
+ * Mutable concurrently accessible data buffer. Could be accessed from several workers simulteniously.
+ */
+@Frozen
+public class MutableData constructor(capacity: Int = 16) {
+    init {
+        if (capacity <= 0) throw IllegalArgumentException()
+        // Instance of MutableData is shared.
+        share()
+    }
+
+    private var buffer = ByteArray(capacity).apply { share() }
+    private var size_ = 0
+    private val lock = Lock()
+
+    private fun resizeDataLocked(newSize: Int): Int {
+        assert(newSize >= size)
+        if (newSize > buffer.size) {
+            val actualSize = maxOf(buffer.size * 3 / 2 + 1, newSize)
+            val newBuffer = ByteArray(actualSize)
+            buffer.copyRangeTo(newBuffer, 0, size, 0)
+            newBuffer.share()
+            buffer = newBuffer
+        }
+        val position = size
+        size_ = newSize
+        return position
+    }
+
+    /**
+     * Current data size, may concurrently change later on.
+     */
+    public val size: Int
+        get() = size_
+
+    /**
+     * Reset the data buffer, makings its size 0.
+     */
+    public fun reset() = locked(lock) {
+        size_ = 0
+    }
+
+    /**
+     * Appends data to the buffer.
+     */
+    public fun append(data: MutableData) = locked(lock) {
+        val toCopy = data.size
+        val where = resizeDataLocked(size + toCopy)
+        data.copyInto(buffer, 0, toCopy, where)
+    }
+
+    /**
+     * Appends byte array to the buffer.
+     */
+    public fun append(data: ByteArray, fromIndex: Int = 0, toIndex: Int = data.size): Unit = locked(lock) {
+        if (toIndex <= fromIndex) return
```

Comment:
I'd expect the contract of `Appendable.append` here, so that IOOBE is thrown when from > to